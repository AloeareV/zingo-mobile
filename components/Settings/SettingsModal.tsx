/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import { View, ScrollView, SafeAreaView, Image, TouchableOpacity } from 'react-native';
import RegText from './Components/RegText';
import FadeText from './Components/FadeText';
import BoldText from './Components/BoldText';
import RegTextInput from './Components/RegTextInput';
import ZecAmount from './Components/ZecAmount';
import { parseServerURI, SERVER_URI, MEMOS } from '../app/uris';
import Button from './Button';
import { useTheme } from '@react-navigation/native';
import { WalletSettings } from '../app/AppState';
//import {TouchableOpacity} from 'react-native-gesture-handler';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faDotCircle } from '@fortawesome/free-solid-svg-icons';
import { faCircle as farCircle } from '@fortawesome/free-regular-svg-icons';

type SettingsProps = {
  closeModal: () => void;
  wallet_settings: WalletSettings;
  set_wallet_option: (name: string, value: string) => void;
  set_server_option: (server: string) => void;
  totalBalance: object;
  currencyName: string;
};

const Settings: React.FunctionComponent<SettingsProps> = ({
  wallet_settings,
  set_wallet_option,
  set_server_option,
  closeModal,
  totalBalance,
  currencyName,
}) => {
  const { colors } = useTheme();

  const [memos, setMemos] = React.useState(wallet_settings.download_memos);
  const [filter, setFilter] = React.useState(wallet_settings.transaction_filter_threshold);
  const [server, setServer] = React.useState(wallet_settings.server);
  const [error, setError] = React.useState(null);

  const [customIcon, setCustomIcon] = React.useState(null);

  React.useEffect(() => {
    setCustomIcon(SERVER_URI.find(s => s === server) ? farCircle : faDotCircle);
  }, [wallet_settings, server]);

  const saveSettings = async () => {
    if (
      wallet_settings.download_memos === memos &&
      wallet_settings.server === server &&
      wallet_settings.transaction_filter_threshold === filter
    ) {
      setError('No changes registred.');
      setTimeout(() => {
        setError(null);
      }, 5000);
      return;
    }
    if (!memos) {
      setError('You need to choose the download memos option to save.');
      setTimeout(() => {
        setError(null);
      }, 5000);
      return;
    }
    if (!filter) {
      setError('You need to choose the transaction filter threshold option to save.');
      setTimeout(() => {
        setError(null);
      }, 5000);
      return;
    }
    if (!server) {
      setError('You need to choose one Server of the list or fill out a valid Server URI.');
      setTimeout(() => {
        setError(null);
      }, 5000);
      return;
    }
    const result = parseServerURI(server);
    if (result.toLowerCase().startsWith('error')) {
      setError('You need to fill out a valid Server URI.');
      setTimeout(() => {
        setError(null);
      }, 5000);
      return;
    }

    if (wallet_settings.download_memos !== memos) {
      set_wallet_option('download_memos', memos);
    }
    if (wallet_settings.transaction_filter_threshold !== filter) {
      set_wallet_option('transaction_filter_threshold', filter);
    }
    if (wallet_settings.server !== server) {
      set_server_option(server);
    }

    closeModal();
  };

  //console.log(wallet_settings);

  return (
    <SafeAreaView
      style={{
        display: 'flex',
        justifyContent: 'flex-start',
        alignItems: 'stretch',
        height: '100%',
        backgroundColor: colors.background,
      }}>
      <View
        style={{
          display: 'flex',
          alignItems: 'center',
          paddingBottom: 10,
          backgroundColor: colors.card,
          zIndex: -1,
          paddingTop: 10,
        }}>
        <Image
          source={require('../assets/img/logobig-zingo.png')}
          style={{ width: 80, height: 80, resizeMode: 'contain' }}
        />
        <ZecAmount currencyName={currencyName} size={36} amtZec={totalBalance.total} style={{ opacity: 0.4 }} />
        <RegText color={colors.money} style={{ marginTop: 5, padding: 5 }}>
          Settings
        </RegText>
        <View style={{ width: '100%', height: 1, backgroundColor: colors.primary }} />
      </View>

      <ScrollView
        style={{ maxHeight: '85%' }}
        contentContainerStyle={{
          flexDirection: 'column',
          alignItems: 'stretch',
          justifyContent: 'flex-start',
          backgroundColor: colors.background,
        }}>
        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>SERVER</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {SERVER_URI.map(uri => (
            <TouchableOpacity
              key={'touch-' + uri}
              style={{ marginRight: 10, marginBottom: 5 }}
              onPress={() => setServer(uri)}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <FontAwesomeIcon icon={uri === server ? faDotCircle : farCircle} size={20} color={colors.border} />
                <RegText key={'tex-' + uri} style={{ marginLeft: 10 }}>
                  {uri}
                </RegText>
              </View>
            </TouchableOpacity>
          ))}

          <View style={{ display: 'flex', flexDirection: 'row' }}>
            <TouchableOpacity style={{ marginRight: 10, marginBottom: 5 }} onPress={() => setServer(null)}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                {customIcon && <FontAwesomeIcon icon={customIcon} size={20} color={colors.border} />}
                <RegText style={{ marginLeft: 10 }}>custom</RegText>
              </View>
            </TouchableOpacity>

            {customIcon === faDotCircle && (
              <RegTextInput
                placeholder={'... http------.---:--- ...'}
                placeholderTextColor={colors.placeholder}
                style={{
                  //flexGrow: 1,
                  fontSize: 18,
                  width: '60%',
                  borderColor: colors.border,
                  borderWidth: 1,
                  marginLeft: 5,
                  padding: 5,
                  paddingTop: 10,
                  paddingBottom: 10,
                  marginTop: Platform.OS === 'ios' ? 15 : 3,
                }}
                value={server}
                onChangeText={(text: string) => setServer(text)}
              />
            )}
          </View>
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>TRANSACTION FILTER THRESHOLD</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          <RegTextInput
            placeholder={'Number...'}
            placeholderTextColor={colors.placeholder}
            keyboardType="numeric"
            style={{
              //flexGrow: 1,
              fontSize: 18,
              width: '60%',
              borderColor: colors.border,
              borderWidth: 1,
              marginLeft: 5,
              padding: 5,
              paddingTop: 10,
              paddingBottom: 10,
              marginTop: Platform.OS === 'ios' ? 15 : 3,
            }}
            value={filter}
            onChangeText={(text: string) => setFilter(text)}
          />
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>MEMO DOWNLOAD</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {MEMOS.map(memo => (
            <View key={'view-' + memo.value}>
              <TouchableOpacity style={{ marginRight: 10, marginBottom: 5 }} onPress={() => setMemos(memo.value)}>
                <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                  <FontAwesomeIcon
                    icon={memo.value === memos ? faDotCircle : farCircle}
                    size={20}
                    color={colors.border}
                  />
                  <RegText key={'text-' + memo.value} style={{ marginLeft: 10 }}>
                    {memo.value}
                  </RegText>
                </View>
              </TouchableOpacity>
              <FadeText key={'fade-' + memo.value}>{memo.text}</FadeText>
            </View>
          ))}
        </View>
      </ScrollView>

      <View
        style={{ flexGrow: 1, flexDirection: 'column', justifyContent: 'center', alignItems: 'center', margin: 20 }}>
        {error && <FadeText style={{ color: colors.primary }}>{error}</FadeText>}
        <View
          style={{ flexGrow: 1, flexDirection: 'row', justifyContent: 'center', alignItems: 'center', marginTop: 5 }}>
          <Button type="Primary" title="Save" style={{ marginLeft: 10 }} onPress={saveSettings} />
          <Button type="Secondary" title="Close" style={{ marginLeft: 10 }} onPress={closeModal} />
        </View>
      </View>
    </SafeAreaView>
  );
};

export default Settings;
