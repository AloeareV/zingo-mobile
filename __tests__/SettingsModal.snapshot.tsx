/**
 * @format
 */

import 'react-native';
import React from 'react';

import { create } from 'react-test-renderer';
import SettingsModal from '../components/SettingsModal';

jest.mock('@fortawesome/react-native-fontawesome', () => ({
    FontAwesomeIcon: ''
}));
jest.useFakeTimers()

// test suite
describe("Component SettingsModal - test", () => {
  //snapshot test
  test("SettingsModal - snapshot", () => {
    const settingsModal = create(<SettingsModal closeModal={() => {}} wallet_settings={{ download_memos: 'wallet' }} set_wallet_option={() => {}} />);
    console.log(settingsModal.toJSON().children[0].children[0].children[2].children[0])
    expect(settingsModal.toJSON()).toMatchSnapshot();
  });

});
