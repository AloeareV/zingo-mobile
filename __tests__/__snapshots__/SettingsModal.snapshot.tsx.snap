// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SettingsModal - test SettingsModal - snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "rgb(242, 242, 242)",
      "display": "flex",
      "height": "100%",
      "justifyContent": "flex-start",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "stretch",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
    style={
      Object {
        "maxHeight": "85%",
      }
    }
  >
    <View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgb(255, 255, 255)",
              "paddingBottom": 25,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(28, 28, 30)",
                "fontSize": 28,
                "marginTop": 5,
                "padding": 5,
              }
            }
          >
            Wallet Settings
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "marginTop": -25,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/logobig.png",
              }
            }
            style={
              Object {
                "height": 50,
                "resizeMode": "contain",
                "width": 50,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "margin": 20,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "rgb(28, 28, 30)",
              },
              Object {
                "fontSize": 18,
              },
              Object {
                "fontWeight": "bold",
              },
              Object {
                "opacity": 0.87,
              },
            ]
          }
        >
          MEMO DOWNLOAD
        </Text>
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "marginLeft": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 10,
              "marginRight": 10,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginTop": 20,
              }
            }
          >
            <
              color="#ffffff"
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f111",
                    "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z",
                  ],
                  "iconName": "circle",
                  "prefix": "far",
                }
              }
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "marginLeft": 10,
                  },
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 18,
                    "fontWeight": "600",
                    "opacity": 0.87,
                  },
                ]
              }
            >
              None
            </Text>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
              "opacity": 0.65,
            }
          }
        >
          Don't download any memos. Server will not learn what transactions belong to you.
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 10,
              "marginRight": 10,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginTop": 20,
              }
            }
          >
            <
              color="#ffffff"
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f192",
                    "M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm80 248c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80z",
                  ],
                  "iconName": "dot-circle",
                  "prefix": "fas",
                }
              }
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "marginLeft": 10,
                  },
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 18,
                    "fontWeight": "600",
                    "opacity": 0.87,
                  },
                ]
              }
            >
              Wallet
            </Text>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
              "opacity": 0.65,
            }
          }
        >
          Download only my memos. Server will learn what TxIDs belong to you, but can't see the addresses, amounts or memos
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 10,
              "marginRight": 10,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "marginTop": 20,
              }
            }
          >
            <
              color="#ffffff"
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f111",
                    "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z",
                  ],
                  "iconName": "circle",
                  "prefix": "far",
                }
              }
              size={20}
            />
            <Text
              style={
                Array [
                  Object {
                    "marginLeft": 10,
                  },
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 18,
                    "fontWeight": "600",
                    "opacity": 0.87,
                  },
                ]
              }
            >
              All
            </Text>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "rgb(28, 28, 30)",
              "opacity": 0.65,
            }
          }
        >
          Download all memos in the blockchain. Server will not learn what TxIDs belong to you. This consumes A LOT of bandwidth.
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "justifyContent": "center",
        "margin": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "rgb(28, 28, 30)",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginLeft": 10,
          "opacity": 1,
          "padding": 10,
          "paddingLeft": 20,
          "paddingRight": 20,
          "width": "40%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(28, 28, 30)",
            "fontWeight": "bold",
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Close
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
